### ISSUE TITLE
Escopo 5 — KV cache: amortized growth instead of full recopy on every append

### ISSUE THREAD

**Issue by fabious054** [fabious054/CandleCLI#11]
## Context

Core item of Escopo 5 (see Trello: "ESCOPO 5 — Otimizar o que já existe" and the design-decision card for capacity/growth). Found during the function-by-function memory allocation audit that also produced issues #8 and #10.

## The problem

`KvCache::append` (`src/arch/qwen/attention.rs`) grows K and V via `Tensor::cat` on every call:

```rust
self.k = Some(match &self.k {
    None => k,
    Some(prev) => Tensor::cat(&[prev, &k], self.dim)?,
});
```

`Tensor::cat` allocates a brand-new tensor sized `previous_len + new_len` and copies the *entire* previous content into it, every single call — not just an append, a full recopy. This runs once per layer (28×) per generated token.

For a conversation with N tokens of history, generating token N+1 costs copying N positions (K and V, each layer). Summed across a whole conversation, total cost is `O(N²)`, not `O(N)` — the classic quadratic KV cache antipattern. This compounds silently under every token generated, on top of whatever prefill cost is already visible.

## What to implement

Replace the `Option<Tensor>` fields with raw `Vec<f32>` buffers that track *capacity* (allocated size) separately from *length* (positions actually filled) — the same growth model `Vec` itself uses, and the same "own the raw buffer, not `candle_core::Tensor`" pattern already used in `src/kernel/matmul.rs` (`candle_core::Tensor` has no concept of reserved-but-unused capacity; it is always exactly the size of its content).

**Design, already decided (see Trello):**
- Initial capacity: 256 positions per layer (~1MB per K, ~1MB per V, per layer — ~56MB total across 28 layers)
- Growth: doubling when capacity is exceeded (256 → 512 → 1024 → ... → 32768) — same amortized-O(1)-per-append behavior as `Vec::push`, ~7 reallocations total across a full conversation up to Qwen3's 32k context ceiling, instead of one recopy per token
- Considered and rejected: starting at 1024 — 4× the memory (~224MB) just to save 1-2 cheap reallocations early in long conversations

**What `append` should do:**
1. If the new slice doesn't fit in remaining capacity, grow the buffer (allocate new capacity, copy only the currently-filled portion — not done on every append, only when capacity is exceeded)
2. Write the new K/V slice into the buffer at the current length offset
3. Advance the filled-length counter
4. Return a view of just the filled portion, wrapped in a `Tensor`, for the rest of `Attention::forward` to consume (repeat_kv, matmul, softmax all still operate on `Tensor` — only the cache's own storage changes, not its public contract)

**What `reset` should do:** drop back to length 0 without necessarily freeing the allocated capacity (a `reset` immediately followed by reuse — e.g. `/reset` in the CLI — shouldn't force paying for a fresh allocation right away).

## Acceptance criteria

- [ ] `KvCache` stores K/V as raw `Vec<f32>` with tracked capacity and length, not `Option<Tensor>`
- [ ] `append` never recopies the whole history — only copies when growing (doubling), and only copies the currently-filled portion when it does
- [ ] `append`'s public signature/behavior unchanged: still takes `&Tensor, &Tensor`, still returns `(Tensor, Tensor)` of the full history so far — `Attention::forward` should need zero changes
- [ ] Existing test `kv_cache_matches_oracle_concat_cache` still passes unchanged
- [ ] New test(s): growth actually happens (a sequence of appends exceeding 256 triggers doubling, verify capacity/behavior), and `reset` followed by reuse doesn't force an immediate reallocation
- [ ] `attention_matches_oracle_equivalent` and `model_matches_oracle_end_to_end` still pass — these exercise `KvCache` indirectly through real attention/model forward passes
- [ ] `cargo test --lib` green
- [ ] No behavior change in output — same logits, same tokens, only how the cache's memory is managed changes

## Documentation (explicit ask from Fabio — always document)

- [ ] Update `docs/DEVELOPMENT.md`'s module structure section to reflect the new `KvCache` design
- [ ] Update `CLAUDE.md`'s equivalent section (Portuguese)
- [ ] Doc comments on the new `KvCache` fields/methods explaining the capacity/length distinction and why (same level of detail as the kernel files from Escopo 4)

## Out of scope

Matmul buffer reuse, attention-without-Tensor-intermediary — separate issues, same Escopo 5.

**Comment by fabious054** [comment]
## Scope revised — merging with attention-without-Tensor

While designing the raw-buffer implementation, found a real technical constraint that changes this issue's scope.

### The finding

A raw `Vec<f32>` buffer with capacity/length tracking eliminates the *reallocation* cost on `append` (no more allocate-a-new-buffer-and-copy-everything on every call). That part of the original design is correct and still holds.

But `append`'s contract (per this issue's original acceptance criteria) still requires returning a `candle_core::Tensor` of the full history for `Attention::forward`'s downstream matmul to consume. Candle's stable API has no way to hand off a `Vec` as `Tensor` storage without copying unless the `Vec`'s length exactly matches the tensor's shape — and our buffer intentionally carries spare capacity beyond the used length, so it can never be handed off directly. That means **materializing the Tensor still copies the entire current history on every single call**, regardless of the storage layer underneath. The `O(N²)` total copy volume across a conversation does not go away with the storage change alone — only the allocation overhead does.

The only way to actually eliminate the repeated full-history copy is for the history to never become a `candle_core::Tensor` mid-pipeline at all — i.e., `repeat_kv`, the QK^T matmul, and softmax also need to operate on the raw buffer directly. That is exactly the "attention without Tensor intermediary" item that was scoped as separate, follow-up work. The two are not actually separable if the goal is to fix the `O(N²)` problem for real, not just reduce allocation churn.

### Decision (Fabio)

Merge this issue's scope with the attention-without-Tensor item. One combined piece of work: `KvCache` as a raw buffer *and* `Attention::forward`'s `repeat_kv`/matmul/softmax working against that buffer directly, only producing a `Tensor` where a `Tensor` is genuinely unavoidable (the final context output handed to `o_proj`, which already goes through `Q8_0Matmul::forward` taking a `Tensor`).

This issue's title and acceptance criteria will be updated to reflect the combined scope before implementation starts. The Escopo 5 Trello card is already updated.

**Comment by fabious054** [comment]
ADR-0005 committed (`de13b06`, branch `escopo5/kv-cache-amortized-growth`): [`docs/adr/ADR-0005-kv-cache-attention-buffer-unification.md`](https://github.com/fabious054/CandleCLI/blob/escopo5/kv-cache-amortized-growth/docs/adr/ADR-0005-kv-cache-attention-buffer-unification.md) — full record of the finding, the two options considered, and why Option A (merge scope) was chosen over Option B (ship the partial storage-only fix). This is now the canonical reference for this issue's revised scope, since the title/acceptance criteria above can't be edited in place (no rename tool available).

**Comment by fabious054** [comment]
## Implemented — combined scope (KV cache + attention without Tensor, per ADR-0005)

Branch `escopo5/kv-cache-amortized-growth`, 7 commits total (including ADR-0005 from the earlier comment):

- `de13b06` — ADR-0005
- `93e9bc1` — `src/kernel/attention.rs` (new): `RawKvCache`, `causal_attention`, tests
- `893bb7d` — `kernel/mod.rs`: export the new types
- `693ddbd` — `arch/qwen/attention.rs`: `Attention::forward` rewritten to use them; old `KvCache` removed
- `e4857bb` — `arch/qwen/layers.rs`: `Layer` uses `RawKvCache` via `attention.new_cache()`
- `0f1117d`, `7230838` — `docs/DEVELOPMENT.md` and `CLAUDE.md` updated, same branch (no separate docs PR)

### What changed, concretely

- **`RawKvCache`**: raw `Vec<f32>` buffer, position-major `(capacity, num_kv_heads, head_dim)`, capacity/length tracked separately. Initial capacity 256, doubles on overflow. `append` copies only the new slice most of the time; growth copies only the currently-filled prefix, not on every call.
- **`causal_attention`**: GQA, causal masking, and softmax computed directly on `f32` slices — no `repeat_kv` materialization (GQA reads KV head `h / n_rep` directly), no mask tensor (the visible-key range is a bounded loop, `offset + i + 1` — a masked-out score is never computed at all), no softmax tensor.
- **`Attention::forward`**: `Tensor` only touches the Q/K/V/O projections (`Q8_0Matmul`, unchanged from Escopo 4), RoPE (deliberately left as-is — `O(new tokens)` only, never the growing history, so it was never the `O(N²)` cost), and wrapping the final context vector for `o_proj`. Everything in between is raw buffers now.
- **`RotaryEmbedding`**: untouched from the Escopo 5 quick-win (issue #8).

### Validation

**New tests** (`kernel::attention`):
- `raw_kv_cache_grows_and_preserves_data` — forces a doubling (256→512) mid-sequence, verifies no data corruption across the growth
- `raw_kv_cache_reset_keeps_capacity` — `/reset` doesn't force an immediate reallocation
- `causal_attention_matches_tensor_reference` — the raw math validated against an independent `Tensor`-based reference (repeat_kv broadcast + explicit mask + `candle_nn`'s softmax) on synthetic data, across a prefill + decode sequence — isolates correctness of the rewrite itself from real model weights, same spirit as Escopo 3's piece-by-piece oracle validation

**Existing tests, unaffected in behavior**:
- `attention_matches_oracle_equivalent` — treats `Attention::forward` as a black box; only had to change how the test constructs a cache (`ours.new_cache()` instead of the removed `KvCache::new(2)`). Still validates the whole rewrite against a true independent oracle on real GGUF weights, at the existing 0.02 threshold (ADR-0004) — no new precision divergence expected, this is the same arithmetic, just off `Tensor`.
- `model_matches_oracle_end_to_end` — exercises a real prefill + decode through the full 28-layer model; unaffected by this change's internals.
- `rope_matches_oracle` — untouched.

**Removed**: `kv_cache_matches_oracle_concat_cache` — tested the old `Tensor`-based `KvCache`'s API directly, which no longer exists. Replaced by the more targeted `RawKvCache` tests above.

### To run

```
git fetch origin
git checkout escopo5/kv-cache-amortized-growth
git pull origin escopo5/kv-cache-amortized-growth
cargo test --lib
```

Expect the same 26 tests as before minus `kv_cache_matches_oracle_concat_cache` plus the 3 new ones in `kernel::attention` = **29 total**. No precision divergence expected anywhere.

After tests pass: `cargo run --release`, a longer multi-turn conversation (the kind that made the second-turn slowdown visible before) is the most informative live test here — watch both `▸ N.N tok/s` and how it holds up as the conversation grows, not just a single prompt.

**Comment by fabious054** [comment]
## Second review pass, before opening the PR — found real waste in my own first rewrite

Looking for anywhere else memory allocation/copying could improve, on the code just committed — found two real problems introduced by the first pass, not pre-existing ones.

### What was wrong

1. **Orphaned `reshape`/`transpose` on q/k/v.** The first version of `Attention::forward` still transposed q/k/v into the old heads-major `Tensor` layout (leftover from the pre-rewrite design, which needed that layout for `repeat_kv`/`Tensor` matmul), then transposed them *back* to position-major before flattening for the raw buffer. Two full round-trip copies per tensor, for nothing — `causal_attention` never wanted the heads-major layout in the first place. For `v` specifically (never touched by norm or RoPE) that transpose served no purpose at all.
2. **Mixing raw and `Tensor` mid-pipeline.** `q_norm`/`k_norm` and RoPE were still `Tensor`-based. Once the KV cache and attention math were raw, that meant bouncing raw → `Tensor` → raw right in the middle — strictly worse than the original all-`Tensor` version, not an improvement.

### The fix

- `Q8_0Matmul::raw_forward` called directly on `q`/`k`/`v` projections — output is already row-major `(l, heads, head_dim)`, no reshape/transpose needed at all.
- `kernel/rmsnorm.rs` (`rms_norm_rows`) — RMSNorm's core moved off `Tensor`. `RmsNorm::forward` (used by `attn_norm`/`ffn_norm`/`output_norm`) keeps its `Tensor` signature unchanged but now does one flatten + one rebuild internally instead of ~4 tensor allocations. New `forward_raw` skips `Tensor` entirely for `q_norm`/`k_norm`.
- `kernel/rope.rs` (`RawRotaryEmbedding`) — replaces the old `Tensor`-based `RotaryEmbedding` entirely. Not because RoPE was ever the `O(N²)` cost (it only ever touches new tokens), but because leaving it on `Tensor` once everything around it was raw would recreate the mid-pipeline seam problem above.

`Attention::forward` now touches `Tensor` exactly twice: flattening `x` on the way in, wrapping the context vector for `o_proj` on the way out. Both `O(new tokens)`, never `O(history)`.

### New tests

- `kernel::rmsnorm`: `rms_norm_matches_hand_computed_reference`, `rms_norm_rows_are_independent`
- `kernel::rope`: `rope_matches_oracle` (moved from `arch/qwen/attention.rs`, adapted for the raw row-major layout vs. the oracle's `Tensor`-based one)

`attention_matches_oracle_equivalent` (unchanged in what it tests, just constructs `RawRotaryEmbedding` now) and `model_matches_oracle_end_to_end` both still validate the whole rewrite as black boxes against the real oracle — no new precision divergence expected anywhere, this is the same arithmetic throughout, just off `Tensor`.

Documentation (`DEVELOPMENT.md`, `CLAUDE.md`) updated again in the same branch to reflect the final state — the first docs commit on this branch had gotten out of sync with the code by the time this second pass landed, caught and fixed before the PR.

### Expected test count

28 (your last run) + 2 (rmsnorm) + 1 (new rope location) − 1 (old rope location removed) = **30 total**.

### To run

```
git fetch origin
git checkout escopo5/kv-cache-amortized-growth
git pull origin escopo5/kv-cache-amortized-growth
cargo test --lib
```

Then the same live multi-turn test as before — this pass should show up as a further tok/s improvement on top of what you already measured (16.4 → 12.6 → 11.8), since real per-token work (not just allocation count) went down.
